executor/index_advise.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"context"
+	"strings"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser"
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+// IndexAdviseExec represents a index advise executor.
+type IndexAdviseExec struct {
+	baseExecutor
+
+	IsLocal         bool
+	indexAdviseInfo *IndexAdviseInfo
+}
+
+// Next implements the Executor Next interface.
+func (e *IndexAdviseExec) Next(ctx context.Context, req *chunk.Chunk) error {
+	if !e.IsLocal {
+		return errors.New("Index Advise: don't support load data without local field")
+	}
+	if e.indexAdviseInfo.Path == "" {
+		return errors.New("Index Advise: infile path is empty")
+	}
+
+	if val := e.ctx.Value(IndexAdviseVarKey); val != nil {
+		e.ctx.SetValue(IndexAdviseVarKey, nil)
```

Comment:
We should not set key to nil here? otherwise, the next `advise index` query may get chance to run while the first `advise index` has not finished yet.

Reply:
No, it has to be cleared. In a session, you could not execute another sql until previous one has finished, so if it has key from previous execution, it must be the case that it meets errors and we failed to remove it.

Reply:
I'm not very clear about this part. We seem to finish the `index advise` process in the optimizer, but we will not execute the SQL in the workload file. 
Sorry i just ref the *load data* before but do not consider it clearly. 😶Can I get more details? or just give me some code link for this problem.

Reply:
To understand this problem, you need to know the workflow of sql. Normally sql would finish most of their jobs after a call of `Execute` in https://github.com/pingcap/tidb/pull/13968/files#diff-dc417f5322100e2e83a8818c70a25878R1186, and you can infer from the name that it would do optimizer and init executors etc. Where `Next` would be called depends on if it is a no no delay executor(check the function `(a *ExecStmt) handleNoDelay)`, if so, `Next` is finshed before the return of `Execute`, otherwise `Next` is only called after `Execute` when write the result set.

Reply:
I think I know the workflow, `Next` of `IndexAdviseExec` will be finished directly. 
I guess i just thought of if there are some `index advise` SQLs in workload file, we may get multiple `IndexAdviseExec` in one session (I forgot we don't have to execute the SQLs 🙃), then I throw the error: `there is already an IndexAdviseExec running in session`, which confuses @eurekaka( i'm so sorry 😅). 
Both of you are right. Now I know how to fix it, i need to let db throw a correct error. Thanks a lot.  @lamxTyler @eurekaka